### PR TITLE
Add/cnn example

### PR DESCRIPTION
**Pull request by jlnav**

### CODE REVIEW

**Comment by shuds13** [comment]
That is the way I see it.

The training data is loaded in once at the beginning of the SIM where each worker gets its chunk of the data.  A simple block decomposition.

In sim:

```python

   for epoch in range(epochs):
        for batch in batches:  # lets start with single batch
            # optimizer.zero_grad() # done on gen
            output = model(data)
            loss = nn.CrossEntropyLoss()(output, target)
            loss.backward()
            libE send_recv (send gradients get parameters)

            # optimizer.step() done on gen
```

In gen:

```python
optimizer.zero_grad() 
Receive and combine gradients + attach to optimizer somehow.
optimizer.step() 
Send parameters.
```


I'm a bit concerned about having the optimizer on the Gen  in case of any changes outside of parameters but if we take this approach then this is how it works

1. Let's get it working first without using batches (all a single batch).

2. For batches:
If we are dividing our dataset into batches, where the parameters are updated at the end of each batch. As the data being processed on the different workers will be happening with the original model parameters, a single batch would consist of the  sum of the local batches being processed across all the workers in any one cycle. I guess this works, and that the order of the "local" batches could be shuffled on each worker (making it stochastic!). The generator has to be called after each batch.

**Comment by shuds13** [comment]
Also I think this is a good example of using proxy store

**Review by jmlarson1 — APPROVED** [review]

File: gen_updates_cnn/README.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+and unloaded. Data-parallel or cross-node techniques in many AI libraries are difficult
+to run on multi-node systems, or assume user familiarity with parallel programming.
+
+libEnsemble allows easy parallelization across multiple nodes and HPC
+systems. "Ensembles" of experiments with libEnsemble take no configuration
+to run on separate nodes and intercommunicate. 
+
+In this example, one model is optimized in parallel by a parent "generator"
+model based on the summed gradients from the N "simulator" models. For this
+simpler example, users must only specify the number of workers `-n N` and start
```
Comment (shuds13) [inline gen_updates_cnn/README.md:29]: simple

File: gen_updates_cnn/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+If the local dataset for training a single model must remain small,
+then time-to-generalization during training may be high as batches must be loaded
+and unloaded. Data-parallel or cross-node techniques in many AI libraries are difficult
+to run on multi-node systems, or assume user familiarity with parallel programming.
+
+libEnsemble allows easy parallelization across multiple nodes and HPC
+systems. "Ensembles" of experiments with libEnsemble take no configuration
+to run on separate nodes and intercommunicate. 
+
+In this example, one model is optimized in parallel by a parent "generator"
```
Comment (shuds13) [inline gen_updates_cnn/README.md:27]: I don't quite get this sentence ‘optimized in parallel by a parent "generator"’. It's been done by the generator than that part is not in parallel.

File: gen_updates_cnn/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+of the dataset and send their gradients to a generator process. The generator process
+combines the gradients from all workers, updates the model parameters, and sends 
+the updated model back to the workers.
+
+Given the number of gradients/parameters that are being updated, `proxystore`
+is used.
+
+## Justification
+
+If the local dataset for training a single model must remain small,
```
Comment (shuds13) [inline gen_updates_cnn/README.md:18]: Rather than remains small I would emphasise must fit in memory

File: gen_updates_cnn/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+### Local setup
+
+Start a redis server instance to hold streaming data:
+
+`redis-server`
+
+### Multi-node
+
+TODO
```
Comment (shuds13) [inline gen_updates_cnn/README.md:55]: TODO

File: gen_updates_cnn/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+## Simulator
+
+Runs model training code without optimization. Sends gradients across
```
Comment (shuds13) [inline gen_updates_cnn/README.md:59]: Maybe say computes the loss on back propagation?

File: gen_updates_cnn/run_libe_cnn.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+if __name__ == "__main__":
+
+    ensemble = Ensemble(parse_args=True)
+
+    ensemble.libE_specs.gen_on_manager = True
+
+    user = {"num_networks": ensemble.nworkers}
+
+    sim_specs = SimSpecs(sim_f=mnist_training_sim, user=user)
+    sim_specs.user["max_epochs"] = 1
```
Comment (shuds13) [inline gen_updates_cnn/run_libe_cnn.py:18]: Maybe make 2

File: gen_updates_cnn/gen_updates_cnn/utils.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    else:
+        device = torch.device("cpu")
+    return device
+
+
+def _get_datasets(worker_id, num_networks):
+    """Get datasets for training and testing, splitting into even chunks for each worker"""
+    # TODO: refactor this
+
+    train_kwargs = {"batch_size": 640, "shuffle": True}
```
Comment (shuds13) [inline gen_updates_cnn/gen_updates_cnn/utils.py:33]: I would make default batch_size bigger.

**Review by shuds13 — APPROVED** [review]
Some minor suggestions. Approve.
